rust/arrow/src/array/array_string.rs

```diff
@@ -205,28 +205,30 @@ where
         let (_, data_len) = iter.size_hint();
         let data_len = data_len.expect("Iterator must be sized"); // panic if no upper bound.
```

Comment:
shouldn't we just remove this panic here and use the lower bound instead? Seems feasible after this change (and before?), and IMO is how it should be (i.e. pre-allocate the lower, not upper, bound).

Reply:
I was thinking that as well, but I think this is still a useful check because (1) it allows the allocation of a more useful initial capacity - the lower iterator bound is often 0; and also (2) this check can easily be satisfied by using `take()`